src/electionguard/discrete_log.py

```diff
@@ -0,0 +1,100 @@
+# pylint: disable=global-statement
+# support for computing discrete logs, with a cache so they're never recomputed
+
+import asyncio
+from typing import Dict, Tuple
+
+from electionguard.singleton import Singleton
+
+from .group import G, ElementModP, ONE_MOD_P, mult_p, int_to_p_unchecked
+
+DLOG_CACHE = Dict[ElementModP, int]
+DLOG_MAX = 1_000_000
+
+
+def discrete_log(element: ElementModP, cache: DLOG_CACHE) -> Tuple[int, DLOG_CACHE]:
+    """
+    Computes the discrete log (base g, mod p) of the given element,
+    with internal caching of results. Should run efficiently when called
+    multiple times when the exponent is at most in the single-digit millions.
+    Performance will degrade if it's much larger.
+
+    For the best possible performance,
+    pre-compute the discrete log of a number you expect to have the biggest
+    exponent you'll ever see. After that, the cache will be fully loaded,
+    and every call will be nothing more than a dictionary lookup.
+    """
+
+    if element in cache:
+        return (cache[element], cache)
+
+    cache = compute_discrete_log_cache(element, cache)
+    return (cache[element], cache)
+
+
+async def discrete_log_async(
+    element: ElementModP,
+    cache: DLOG_CACHE,
+    mutex: asyncio.Lock = asyncio.Lock(),
+) -> Tuple[int, DLOG_CACHE]:
+    """
+    Computes the discrete log (base g, mod p) of the given element,
+    with internal caching of results. Should run efficiently when called
+    multiple times when the exponent is at most in the single-digit millions.
+    Performance will degrade if it's much larger.
+
+    Note: *this function is thread-safe*. For the best possible performance,
+    pre-compute the discrete log of a number you expect to have the biggest
+    exponent you'll ever see. After that, the cache will be fully loaded,
+    and every call will be nothing more than a dictionary lookup.
+    """
+    if element in cache:
+        return (cache[element], cache)
+
+    async with mutex:
+        if element in cache:
+            return (cache[element], cache)
+
+        cache = compute_discrete_log_cache(element, cache)
+        return (cache[element], cache)
+
+
+def compute_discrete_log_cache(
+    element: ElementModP, cache: DLOG_CACHE
+) -> Dict[ElementModP, int]:
+    """
+    Compute a discrete log cache up to the specified element.
+    """
+    if not cache:
+        cache = {ONE_MOD_P: 0}
+    max_element = list(cache)[-1]
+    exponent = cache[max_element]
+
+    g = int_to_p_unchecked(G)
+    while element != max_element:
+        exponent = exponent + 1
+        if exponent > DLOG_MAX:
+            raise Exception("size is larger than max.")
```

Comment:
This should be throwing a `ValueError` . 

https://docs.python.org/3/library/exceptions.html#ValueError

Possibly a KeyError or IndexError but not a generic Exception.

Reply:
Or, see comments above, the function should return `Optional[int]` rather than `int` with the possibility of an unchecked exception blowing things up downstream.

Reply:
Also, the error message needs to be more descriptive, like "discrete_log failure; possibly caused by bad data encryption".